src/Processors/Transforms/DeduplicationTokenTransforms.h

```diff
@@ -0,0 +1,236 @@
+#pragma once
+
+#include <Processors/Chunk.h>
+#include <Processors/ISimpleTransform.h>
+
+#include <base/defines.h>
+#include "Common/Logger.h"
+
+
+namespace DB
+{
+    class RestoreChunkInfosTransform : public ISimpleTransform
+    {
+    public:
+        RestoreChunkInfosTransform(Chunk::ChunkInfoCollection chunk_infos_, const Block & header_)
+                : ISimpleTransform(header_, header_, true)
+                , chunk_infos(std::move(chunk_infos_))
+        {}
+
+        String getName() const override { return "RestoreChunkInfosTransform"; }
+
+        void transform(Chunk & chunk) override;
+
+    private:
+        Chunk::ChunkInfoCollection chunk_infos;
+    };
+
+
+namespace DeduplicationToken
+{
+    class TokenInfo : public ChunkInfoCloneable<TokenInfo>
+    {
+    public:
+        TokenInfo() = default;
+        TokenInfo(const TokenInfo & other) = default;
+
+        String getToken() const;
+        String debugToken() const;
+
+        bool empty() const { return parts.empty(); }
+
+        bool isDefined() const { return stage == DEFINED; }
+
+        void addChunkHash(String part);
+        void defineSourceWithChunkHashes();
+
+        void setUserToken(const String & token);
+        void defineSourceWithUserToken(size_t block_number);
+
+        void setViewID(const String & id);
+        void defineViewID(size_t block_number);
```

Comment:
`set` and `define` words are quite similar in meaning, so when you read these methods names, you cannot tell what is the difference. May be worth to rename

Reply:
renamed those methods